config/src/credentials.rs

```diff
@@ -37,26 +41,152 @@ impl Credentials {
         }
     }
 
-    pub fn get_authentication_type(&self) -> String {
-        self.method.clone()
-    }
+    pub async fn keycloak(path: &str) -> Self {
+        // read keycloak json
+        let file = File::open(path).expect("open keycloak.json");
+        let mut buf_reader = BufReader::new(file);
+        let mut buffer = Vec::new();
+        buf_reader.read_to_end(&mut buffer).expect("read to the end");
+
+        // decode json string to struct
+        let key_cloak_json: KeyCloakJson = serde_json::from_slice(&buffer).expect("decode slice to struct");
+
+        // first POST request for access token
+        let access_token = obtain_access_token(
+            &key_cloak_json.auth_server_url,
+            &key_cloak_json.realm,
+            &key_cloak_json.resource,
+            &key_cloak_json.credentials.secret,
+        )
+        .await
+        .expect("obtain access token");
 
-    pub fn get_authentication_token(&self) -> String {
-        self.token.clone()
+        // second POST request for rpt
+        let rpt = authorize(
+            &key_cloak_json.auth_server_url,
+            &key_cloak_json.realm,
+            &access_token,
+        )
+        .await
+        .expect("get rpt");
+        Credentials {
+            method: BEARER.to_owned(),
+            token: rpt,
+        }
     }
 
     pub fn get_request_metadata(&self) -> String {
         format!("{} {}", self.method, self.token)
     }
 }
 
+#[derive(Serialize, Deserialize)]
+struct KeyCloakJson {
+    realm: String,
+    #[serde(rename(deserialize = "bearer-only"))]
```

Comment:
bearer-only, ssl_required, confidential_port , public_client are not needed to be parsed I think.   You really just need the auth-server-url, credentials and resource(aka client_id)